### PR TITLE
Fix problems with rate limiting peers that have existing channels 

### PR DESCRIPTION
**Pull request by remyers**
Various fixes and improvements to the 'PendingChannelsRateLimiter' actor in response to issue #2686 .

### CODE REVIEW

**Comment by remyers** [comment]
Suggested fix from @t-bast: "In the handlers for ReplaceChannelId and RemoveChannelId, we only clean up the pendingPrivateNodeChannels if we didn't find the node in the pendingPublicNodeChannels -> that's incorrect, a node_id may be in both maps, so we need to clean up both in all cases" is addressed by [0edbe7d](https://github.com/ACINQ/eclair/pull/2687/commits/0edbe7d2f033720050008171f42a3c1d7155f00c).

Before this fix, the test I added would fail because the pending requests a node made while private would not be removed/replaced if the node also had pending requests added as a public node.

**Comment by remyers** [comment]
Suggested fix : "In restoring, we register to events only after we've analyzed every channel: this is potentially too late, that means we will miss events fired while we were in restoring: we should register to these events first, and stash the ones that fire while we're restoring" is addressed by [b3fdbaa](https://github.com/ACINQ/eclair/pull/2687/commits/b3fdbaa907951b201c91dc7135dc7806094c685e).

I also found/fixed that we were not filtering out (potentially newly) whitelisted channels during restore.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala [hunk 1]
```diff
@@ -59,34 +64,59 @@ private class PendingChannelsRateLimiter(nodeParams: NodeParams, router: ActorRe
...
+        case Some(pendingChannels) =>
+          context.log.debug(s"removed pending channel $channelId for node $remoteNodeId")
+          val pendingChannels1 = pendingChannels.filterNot(_ == channelId)
+          if (pendingChannels1.isEmpty) {
+            pendingNodeChannels - remoteNodeId
+          } else {
+            pendingNodeChannels + (remoteNodeId -> pendingChannels1)
+          }
+        case _ => pendingNodeChannels
+      }
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:119]: nit: those functions can be moved outside of the `registering` function since they are pure, it's a bit easier to read and more efficient

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala [hunk 2]
```diff
@@ -45,6 +49,18 @@ class PendingChannelsRateLimiterSpec extends ScalaTestWithActorTestKit(ConfigFac
   val newChannelIdPrivate1: ByteVector32 = ByteVector32(hex"077777770000000000000000000000000000000000000000000000000000000")
   val channelIdAtLimit1: ByteVector32 = ByteVector32(hex"0888888880000000000000000000000000000000000000000000000000000000")
   val channelIdAtLimit2: ByteVector32 = ByteVector32(hex"0999999990000000000000000000000000000000000000000000000000000000")
+  val waitingForSigs: InteractiveTxSigningSession.WaitingForSigs = {
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:52]: Honestly it seems a bit overkill to go through the pain of creating a dummy `DATA_WAIT_FOR_DUAL_FUNDING_SIGNED`, I think we're fine just using the other states that are easier to "mock" in tests.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala [hunk 3]
```diff
@@ -59,34 +64,59 @@ private class PendingChannelsRateLimiter(nodeParams: NodeParams, router: ActorRe
...
 
-  private def restoring(channels: Map[PublicKey, Seq[PersistentChannelData]], pendingPublicNodeChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Map[PublicKey, Seq[ByteVector32]]): Behavior[Command] = {
+  private def restoring(channels: Map[PublicKey, Seq[PersistentChannelData]], pendingPublicNodeChannels: Map[PublicKey, Seq[ByteVector32]], pendingPrivateNodeChannels: Map[PublicKey, Seq[ByteVector32]], cachedCommands: Seq[Command], sendGetNode: Boolean = true): Behavior[Command] = {
     channels.headOption match {
       case Some((remoteNodeId, pendingChannels)) =>
         val adapter = context.messageAdapter[Router.GetNodeResponse](r => WrappedGetNodeResponse(pendingChannels.head.channelId, r, None))
-        router ! GetNode(adapter, remoteNodeId)
+        if (sendGetNode) router ! GetNode(adapter, remoteNodeId)
         Behaviors.receiveMessagePartial[Command] {
           case AddOrRejectChannel(replyTo, _, _) =>
```
[the remark this answers is not in the document]
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala:73]: It would be a bit dangerous, because that would block the `OpenChannelInterceptor` which would prevent us from opening channels where we are the initiator...I think it's simpler to reason about if we just reject those while we're initializing, our peer can always retry later.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala [hunk 4]
```diff
@@ -39,17 +39,22 @@ object PendingChannelsRateLimiter {
...
-  private[io] def filterPendingChannels(channels: Seq[PersistentChannelData]): Map[PublicKey, Seq[PersistentChannelData]] = {
+  private[io] def filterPendingChannels(nodeParams: NodeParams, channels: Seq[PersistentChannelData]): Map[PublicKey, Seq[PersistentChannelData]] = {
     channels.filter {
-      case _: DATA_WAIT_FOR_FUNDING_CONFIRMED => true
-      case _: DATA_WAIT_FOR_DUAL_FUNDING_SIGNED => true
-      case _: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED => true
-      case _: DATA_WAIT_FOR_CHANNEL_READY => true
-      case _: DATA_WAIT_FOR_DUAL_FUNDING_READY => true
+      case p: PersistentChannelData if nodeParams.channelConf.channelOpenerWhitelist.contains(p.remoteNodeId) => false
+      case DATA_WAIT_FOR_FUNDING_CONFIRMED(commitments, _, _, _) if !commitments.params.localParams.isInitiator => true
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala]: Fixed in [3f71eda](https://github.com/ACINQ/eclair/pull/2687/commits/3f71eda159ac875aa6737564329db1e9f49528c0).

File: eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala [hunk 5]
```diff
@@ -39,17 +39,22 @@ object PendingChannelsRateLimiter {
 
   def apply(nodeParams: NodeParams, router: ActorRef[Router.GetNode], channels: Seq[PersistentChannelData]): Behavior[Command] = {
     Behaviors.setup { context =>
-      new PendingChannelsRateLimiter(nodeParams, router, context).restoring(filterPendingChannels(channels), Map(), Map())
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelIdAssigned](c => ReplaceChannelId(c.remoteNodeId, c.temporaryChannelId, c.channelId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => RemoveChannelId(c.remoteNodeId, c.channelId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelClosed](c => RemoveChannelId(c.commitments.remoteNodeId, c.channelId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelAborted](c => RemoveChannelId(c.remoteNodeId, c.channelId)))
+      new PendingChannelsRateLimiter(nodeParams, router, context).restoring(filterPendingChannels(nodeParams: NodeParams, channels), Map(), Map(), Seq())
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/main/scala/fr/acinq/eclair/io/PendingChannelsRateLimiter.scala]: Fixed in [3f71eda](https://github.com/ACINQ/eclair/pull/2687/commits/3f71eda159ac875aa6737564329db1e9f49528c0).

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala (hunk 2 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala]: Good point; it was a pita to add and it seems unlikely to catch anything extra - at the cost of readability for the test.
Fixed in [3f71eda](https://github.com/ACINQ/eclair/pull/2687/commits/3f71eda159ac875aa6737564329db1e9f49528c0).

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala [hunk 6]
```diff
@@ -295,29 +318,112 @@ class PendingChannelsRateLimiterSpec extends ScalaTestWithActorTestKit(ConfigFac
...
+    import f._
+
+    val events = Seq[ChannelEvent](
+      ChannelIdAssigned(null, peersAtLimit.head, channelIdAtLimit1, newChannelId1),
+      ChannelAborted(null, peersBelowLimit.last, channelIdBelowLimit2),
+      ChannelOpened(null, privatePeers.head, channelIdPrivate1),
+      ChannelOpened(null, peersAtLimit.head, newChannelId1))
+    val limiter = testKit.spawn(PendingChannelsRateLimiter(nodeParams, router.ref, allChannels))
+
+    // trigger the limiter to start the restore
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:412]: What does that mean? The limiter is already in the `restoring` state, isn't it? You're rather testing that while we're restoring, we reject channel requests?

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala [hunk 7]
```diff
@@ -295,29 +318,112 @@ class PendingChannelsRateLimiterSpec extends ScalaTestWithActorTestKit(ConfigFac
...
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, randomKey().publicKey, randomBytes32())
+    probe.expectMessage(PendingChannelsRateLimiter.ChannelRateLimited)
+
+    // publish events that close restored pending channel opens
+    events.foreach(system.eventStream ! Publish(_))
+
+    processRestoredChannnels(f, allChannels)
+    eventually {
+      limiter ! PendingChannelsRateLimiter.CountOpenChannelRequests(requests.ref, publicPeers = true)
+      requests.expectMessage(4)
```
Comment (t-bast) [inline eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:422]: It's a bit hard to know whether this is accurate. It would be easier to understand if you made a first call to `CountOpenChannelRequests` *before* processing restored channels so that readers can compare the values.

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala [hunk 8]
```diff
@@ -295,29 +318,112 @@ class PendingChannelsRateLimiterSpec extends ScalaTestWithActorTestKit(ConfigFac
...
+    // the third pending channel request from peer is accepted as a public peer
+    limiter ! PendingChannelsRateLimiter.AddOrRejectChannel(probe.ref, peer, channelId3)
+    router.expectMessageType[GetNode].replyTo ! PublicNode(announcement(peer), 1, 1 sat)
+    probe.expectMessage(PendingChannelsRateLimiter.AcceptOpenChannel)
+
+    // the private pending channel request from peer receives a new channel id
+    system.eventStream ! Publish(ChannelIdAssigned(null, peer, channelId1, randomBytes32()))
+
+    // the private channel request from peer is aborted and removed from the tracker
+    system.eventStream ! Publish(ChannelAborted(null, peer, channelId2))
```
[the remark this answers is not in the document]
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala]: No,  your fix [3b77778](https://github.com/ACINQ/eclair/pull/2687/commits/3b777789fc0faee7e7d03efc128dd82256a30a0e) makes it properly exercise both the replacement and removing.

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala (hunk 6 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala]: Fixed in [61e3c31](https://github.com/ACINQ/eclair/pull/2687/commits/61e3c3170cc9c62623fc52583c1ea6cd62c23d6f). I need a way to make sure the `limiter` has been created an subscribed to events before I publish the test events. What I had before was a dirty hack, this fix makes it cleaner/clearer.

File: eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala (hunk 7 above)
Comment (remyers) [reply eclair-core/src/test/scala/fr/acinq/eclair/io/PendingChannelsRateLimiterSpec.scala:437]: Fixed in [c7ee927](https://github.com/ACINQ/eclair/pull/2687/commits/c7ee927933048bcc44564dbbab7d9d553252522a)

**Review by t-bast — APPROVED** [review]
I think the tests deserve to be improved, they're quite hard to follow now, but that can be done in a follow-up PR, let's merge this now to unblock the release.
